File: 22_专题/区间和/前缀/积分图integral image/308. 二维区域和检索 - 可变.ts
import { BIT } from '../../../../6_tree/树状数组/BIT'

// 把二维矩阵转化为一维数组，然后使用树状数组
class NumMatrix {
  private m: number
  private n: number
  private matrix: number[][]
  private bit: BIT

  constructor(matrix: number[][]) {
    this.m = matrix.length
    this.n = matrix[0].length
    this.matrix = matrix
    this.bit = new BIT(this.m * this.n)
    for (let i = 0; i < this.m; i++) {
      for (let j = 0; j < this.n; j++) {
        const index = this.getIndex(i, j)
        this.bit.add(index, matrix[i][j])
      }
    }
  }

  update(row: number, col: number, val: number): void {
    const index = this.getIndex(row, col)
    const diff = val - this.matrix[row][col]
    this.bit.add(index, diff)
    this.matrix[row][col] = val // 不要忘了更新原来的矩阵
  }

  sumRegion(row1: number, col1: number, row2: number, col2: number): number {
    let res = 0
    for (let row = row1; row <= row2; row++) {
      const leftIndex = this.getIndex(row, col1)
      const rightIndex = this.getIndex(row, col2)
      res += this.bit.sumRange(leftIndex, rightIndex)
    }
    return res
  }

  private getIndex(row: number, col: number) {
    return this.n * row + col + 1 // 注意这里加一 让树状数组查询索引是正数
  }
}

const test = new NumMatrix([
  [2, 4],
  [-3, 5],
])
test.update(0, 1, 3)
test.update(1, 1, -3)
test.update(0, 1, 1)
console.log(test.sumRegion(0, 0, 1, 1))
// test.update
export {}

File: 6_tree/树状数组/经典题/308. 二维区域和检索 - 可变.py
from typing import List
from BIT import BIT4, BIT5

# 1 <= m, n <= 200
# 最多调用1e4 次 sumRegion 和 update 方法


class NumMatrix:
    def __init__(self, matrix: List[List[int]]):
        self.matrix = matrix
        self.row = len(matrix)
        self.col = len(matrix[0])
        self.bit = BIT4(self.row, self.col)
        for r in range(self.row):
            for c in range(self.col):
                self.bit.add(r, c, matrix[r][c])

    def update(self, row: int, col: int, val: int) -> None:
        """更新 matrix[row][col] 的值到 val 。"""
        delta = val - self.matrix[row][col]
        self.matrix[row][col] = val
        self.bit.add(row, col, delta)

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        """返回矩阵 matrix 中指定矩形区域元素的 和 ，该区域由 左上角 (row1, col1) 和 右下角 (row2, col2) 界定。"""
        return self.bit.queryRange(row1, col1, row2, col2)


class NumMatrix2:
    def __init__(self, matrix: List[List[int]]):
        self.matrix = matrix
        self.row = len(matrix)
        self.col = len(matrix[0])
        self.bit = BIT5(self.row, self.col)
        for r in range(self.row):
            for c in range(self.col):
                self.bit.addRange(r, c, r, c, matrix[r][c])

    def update(self, row: int, col: int, val: int) -> None:
        """更新 matrix[row][col] 的值到 val 。"""
        delta = val - self.matrix[row][col]
        self.matrix[row][col] = val
        self.bit.addRange(row, col, row, col, delta)

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        """返回矩阵 matrix 中指定矩形区域元素的 和 ，该区域由 左上角 (row1, col1) 和 右下角 (row2, col2) 界定。"""
        return self.bit.queryRange(row1, col1, row2, col2)
